larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.h

```diff
@@ -0,0 +1,249 @@
+/**
+ *  @file   larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.h
+ *
+ *  @brief  Header file for the discrete probability helper class.
+ *
+ *  $Log: $
+ */
+#ifndef LAR_DISCRETE_PROBABILITY_HELPER_HELPER_H
+#define LAR_DISCRETE_PROBABILITY_HELPER_HELPER_H 1
+
+#include "larpandoracontent/LArObjects/LArDiscreteProbabilityVector.h"
+
+#include <algorithm>
+#include <random>
+
+namespace lar_content
+{
+
+/**
+ *  @brief  LArDiscreteProbabilityHelper class
+ */
+class LArDiscreteProbabilityHelper
+{
+public:
+    /**
+     *  @brief  Calculate P value for measured correlation coefficient between two datasets via a permutation test 
+     *
+     *  @param  t1 the first input dataset
+     *  @param  t2 the second input dataset
+     *  @param  randomNumberGenerator the random number generator to shuffle the datasets
+     *  @param  nPermutations the number of permutations to run
+     *
+     *  @return the p-value
+     */
+    template <typename T>
+    static float CalculateCorrelationCoefficientPValueFromPermutationTest(const T &t1, const T &t2, 
+        std::mt19937 &randomNumberGenerator, const size_t nPermutations);
+
+    /**
+     *  @brief  Calculate P value for measured correlation coefficient between two datasets via a integrating the student T dist.
+     *
+     *  @param  t1 the first input dataset
+     *  @param  t2 the second input dataset
+     *  @param  nIntegrationSteps how many steps to use in the trapezium integration
+     *
+     *  @return the p-value
+     */
+    template <typename T>
+    static float CalculateCorrelationCoefficientPValueFromStudentTDistribution(const T &t1, const T &t2, 
+        const size_t nIntegrationSteps);
+
+    /**
+     *  @brief  Calculate the correlation coefficient between two datasets 
+     *
+     *  @param  t1 the first input dataset
+     *  @param  t2 the second input dataset
+     *
+     *  @return the correlation coefficient
+     */
+    template <typename T>
+    static float CalculateCorrelationCoefficient(const T &t1, const T &t2);
+
+    /**
+     *  @brief  Calculate the mean of a dataset 
+     *
+     *  @param  t the dataset
+     *
+     *  @return the mean
+     */
+    template <typename T>
+    static float CalculateMean(const T &t);
+
+private:
+    /**
+     *  @brief  Creates a randomised copy of a dataset 
+     *
+     *  @param  t the dataset to be shuffled
+     *  @param  randomNumberGenerator the random number generator
+     *
+     *  @return the reshuffled dataset
+     */
+    template <typename T>
+    static T MakeRandomisedSample(const T &t, std::mt19937 &randomNumberGenerator);
```

Comment:
John and I discussed this one in private.  Passing an empty object to be filled by non-const reference wouldn't work here unfortunately, because one specialisation of this function needs to make a DiscreteProbabilityVector, which cannot be filled after construction (only complete construction of that object is allowed).  I did a bit of digging on Return Value Optimisation (RVO) and the general internet hive mind says that returning by value is the preferred option; the compiler employ RVO to optimise out copies or issue a std::move or whatever else.